File: apps/android/core/src/main/kotlin/app/trackevolution/core/CompareLaps.kt
package app.trackevolution.core

import app.trackevolution.core.model.EventDetail
import app.trackevolution.core.model.Lap
import app.trackevolution.core.model.LapChannels
import app.trackevolution.core.model.Session
import app.trackevolution.core.model.SessionChannels
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/**
 * Cross-event lap comparison (#165): the pure half of "compare two laps" —
 * pick any two laps with stored channel data at a track, put them on one
 * distance grid, and reduce each to head-to-head numbers.
 *
 * A port of `public/js/compare-laps.js` — same function names, same inclusive
 * thresholds — pinned against the web implementation by
 * `contracts/logic/compare-laps.json` ([CompareLapsTest]). The pair
 * [alignLapPair] returns is an ordinary [SessionChannels], so the existing
 * chart stack draws a cross-event pair the same way it draws one session.
 */
public object CompareLaps {

    /**
     * "Flat out" / "on the brakes" thresholds for [lapMetrics]' percentages.
     * Display semantics, not physics — see the JS.
     */
    public const val FULL_THROTTLE_PCT: Double = 95.0
    public const val BRAKING_PCT: Double = 10.0

    /**
     * Two laps whose driven lengths differ by more than this are probably not
     * the same layout / start line — the screen warns past it.
     */
    public const val LENGTH_MISMATCH_WARN: Double = 0.05

    /**
     * One event's contribution to the picker: what [comparableLaps] reads off
     * an event detail. A class of its own (rather than [EventDetail]) so the
     * contract test can build inputs without fabricating a full event row.
     */
    public data class EventLaps(
        val eventId: Int,
        val date: String,
        val club: String?,
        val sessions: List<SessionLaps>,
    ) {
        public constructor(detail: EventDetail) : this(
            eventId = detail.event.id,
            date = detail.event.startDate,
            club = detail.event.club,
            sessions = detail.sessions.map(::SessionLaps),
        )
    }

    public data class SessionLaps(
        val sessionId: Int,
        val label: String?,
        val laps: List<Lap>,
        val channels: SessionChannels?,
    ) {
        public constructor(session: Session) : this(
            sessionId = session.id,
            label = session.label,
            laps = session.laps,
            channels = session.channels,
        )
    }

    /**
     * One pickable lap: a stored lap row that [ChannelGraphs.matchLapsToChannels]
     * paired with a channel entry. [chIdx] indexes that session's `channels.laps`.
     */
    public data class Row(
        val eventId: Int,
        val date: String,
        val club: String?,
        val sessionId: Int,
        val sessionLabel: String?,
        val lapNum: Int,
        val timeMs: Int,
        val chIdx: Int,
    )

    /** Indexes into the rows [defaultComparePicks] was given. */
    public data class Picks(val a: Int, val b: Int)

    /**
     * Flatten event details into one pickable list of laps that have channel
     * data. Keeps the given event order; within an event, session order then
     * lap order. Laps without a channel entry (hand-added, no distance window)
     * drop out, exactly as on the web.
     */
    public fun comparableLaps(events: List<EventLaps>): List<Row> {
        val rows = mutableListOf<Row>()
        for (event in events) {
            for (session in event.sessions) {
                val channels = session.channels ?: continue
                if (channels.laps.isEmpty() || session.laps.isEmpty()) continue
                for (match in ChannelGraphs.matchLapsToChannels(session.laps, channels.laps)) {
                    if (!match.hasChannels) continue
                    rows +=
                        Row(
                            eventId = event.eventId,
                            date = event.date,
                            club = event.club,
                            sessionId = session.sessionId,
                            sessionLabel = session.label,
                            lapNum = match.lap.lapNum,
                            timeMs = match.lap.timeMs,
                            chIdx = match.chIdx,
                        )
                }
            }
        }
        return rows
    }

    /**
     * Default selection: side A is "current me" — the best lap of the most
     * recent event with comparable laps — and side B is "best me", the overall
     * best lap; when they coincide, B falls back to the best of the rest.
     * Null when there is nothing to compare.
     */
    public fun defaultComparePicks(rows: List<Row>): Picks? {
        if (rows.size < 2) return null
        val latest = rows.maxOf { it.date }
        fun bestIn(idxs: List<Int>): Int = idxs.reduce { m, i -> if (rows[i].timeMs < rows[m].timeMs) i else m }
        val all = rows.indices.toList()
        val a = bestIn(all.filter { rows[it].date == latest })
        var b = bestIn(all)
        if (b == a) b = bestIn(all.filter { it != a })
        return Picks(a, b)
    }

    /**
     * Linear-resample one stored channel-lap entry from its grid spacing onto
     * another. Identity when the spacings already match (every writer uses
     * 20 m today, but `dStepM` is stored per session and this must not
     * assume). Same arithmetic as the JS, so the fixture pins every value.
     */
    public fun resampleChannelLap(entry: LapChannels, fromStepM: Double, toStepM: Double): LapChannels {
        if (fromStepM == toStepM) return entry
        fun resample(arr: List<Double>?): List<Double>? {
            if (arr == null || arr.size < 2) return null
            val n = floor((arr.size - 1) * fromStepM / toStepM).toInt() + 1
            return List(n) { k ->
                val p = k * toStepM / fromStepM
                val i0 = min(arr.size - 2, floor(p).toInt())
                arr[i0] + (arr[i0 + 1] - arr[i0]) * (p - i0)
            }
        }
        return LapChannels(
            n = entry.n,
            timeMs = entry.timeMs,
            speed = resample(entry.speed),
            rpm = resample(entry.rpm),
            latG = resample(entry.latG),
            throttle = resample(entry.throttle),
            brake = resample(entry.brake),
            steering = resample(entry.steering),
        )
    }

    /**
     * Put a pair of channel-lap entries on one grid: side B is resampled onto
     * side A's spacing when the two sessions stored different grids.
     */
    public fun alignLapPair(
        entryA: LapChannels,
        stepA: Double,
        entryB: LapChannels,
        stepB: Double,
    ): SessionChannels =
        SessionChannels(v = 1, dStepM = stepA, laps = listOf(entryA, resampleChannelLap(entryB, stepB, stepA)))

    /** Driven length a stored entry covers (its grid extent, meters). */
    public fun drivenLengthM(entry: LapChannels, stepM: Double): Double {
        val n = entry.speed?.size ?: 0
        return if (n > 1) (n - 1) * stepM else 0.0
    }

    /**
     * Relative driven-length difference between two entries (0 = identical).
     * Distance is measured from each lap's own start line, so a large ratio
     * means a different layout, picked line, or off-track excursion — the
     * comparison still renders, with a warning past [LENGTH_MISMATCH_WARN].
     */
    public fun lengthMismatchRatio(
        entryA: LapChannels,
        stepA: Double,
        entryB: LapChannels,
        stepB: Double,
    ): Double {
        val la = drivenLengthM(entryA, stepA)
        val lb = drivenLengthM(entryB, stepB)
        val longest = max(la, lb)
        return if (longest > 0) abs(la - lb) / longest else 0.0
    }

    /**
     * One lap's channels reduced to head-to-head numbers. Speed stays km/h as
     * stored (the screen converts for display); percentages are shares of grid
     * samples — on a uniform distance grid, shares of driven distance.
     * Channels the lap didn't store are null.
     */
    public data class Metrics(
        val timeMs: Int,
        val topSpeedKph: Double?,
        val minSpeedKph: Double?,
        val avgSpeedKph: Double?,
        val maxRpm: Double?,
        val maxLatG: Double?,
        val fullThrottlePct: Double?,
        val brakingPct: Double?,
    )

    public fun lapMetrics(entry: LapChannels): Metrics {
        val speed = entry.speed?.takeIf { it.isNotEmpty() }
        fun share(arr: List<Double>?, minValue: Double): Double? {
            if (arr.isNullOrEmpty()) return null
            return 100.0 * arr.count { it >= minValue } / arr.size
        }
        fun maxOf(arr: List<Double>?): Double? = arr?.takeIf { it.isNotEmpty() }?.max()
        return Metrics(
            timeMs = entry.timeMs,
            topSpeedKph = speed?.max(),
            minSpeedKph = speed?.min(),
            avgSpeedKph = speed?.let { it.sum() / it.size },
            maxRpm = maxOf(entry.rpm),
            maxLatG = maxOf(entry.latG),
            fullThrottlePct = share(entry.throttle, FULL_THROTTLE_PCT),
            brakingPct = share(entry.brake, BRAKING_PCT),
        )
    }
}

File: apps/ios/Packages/TrackEvolutionKit/Sources/TrackEvolutionKit/Charts/CompareLaps.swift
import Foundation

/// Cross-event lap comparison (#165): the pure half of "compare two laps" —
/// pick any two laps with stored channel data at a track, put them on one
/// distance grid, and reduce each to head-to-head numbers.
///
/// A port of `public/js/compare-laps.js` — same function names, same inclusive
/// thresholds — pinned against the web implementation by
/// `contracts/logic/compare-laps.json` (`CompareLapsTests`). The pair that
/// `alignLapPair` returns is an ordinary `SessionChannels`, so the existing
/// chart stack draws a cross-event pair the same way it draws one session.
public enum CompareLaps {
    /// "Flat out" / "on the brakes" thresholds for `lapMetrics`' percentages.
    /// Display semantics, not physics — see the JS.
    public static let FULL_THROTTLE_PCT: Double = 95
    public static let BRAKING_PCT: Double = 10

    /// Two laps whose driven lengths differ by more than this are probably not
    /// the same layout / start line — the screen warns past it.
    public static let LENGTH_MISMATCH_WARN = 0.05

    /// One event's contribution to the picker: what `comparableLaps` reads off
    /// an event detail. A struct of its own (rather than `EventDetail`) so the
    /// contract test can build inputs without fabricating a full event row.
    public struct EventLaps: Sendable {
        public var eventId: Int
        public var date: String
        public var club: String?
        public var sessions: [SessionLaps]

        public init(eventId: Int, date: String, club: String?, sessions: [SessionLaps]) {
            self.eventId = eventId
            self.date = date
            self.club = club
            self.sessions = sessions
        }

        public init(detail: EventDetail) {
            self.init(
                eventId: detail.event.id,
                date: detail.event.startDate,
                club: detail.event.club,
                sessions: detail.sessions.map(SessionLaps.init(session:))
            )
        }
    }

    public struct SessionLaps: Sendable {
        public var sessionId: Int
        public var label: String?
        public var laps: [Lap]
        public var channels: SessionChannels?

        public init(sessionId: Int, label: String?, laps: [Lap], channels: SessionChannels?) {
            self.sessionId = sessionId
            self.label = label
            self.laps = laps
            self.channels = channels
        }

        public init(session: Session) {
            self.init(sessionId: session.id, label: session.label, laps: session.laps, channels: session.channels)
        }
    }

    /// One pickable lap: a stored lap row that `matchLapsToChannels` paired
    /// with a channel entry. `chIdx` indexes that session's `channels.laps`.
    public struct Row: Sendable, Equatable {
        public var eventId: Int
        public var date: String
        public var club: String?
        public var sessionId: Int
        public var sessionLabel: String?
        public var lapNum: Int
        public var timeMs: Int
        public var chIdx: Int

        public init(
            eventId: Int, date: String, club: String?, sessionId: Int, sessionLabel: String?,
            lapNum: Int, timeMs: Int, chIdx: Int
        ) {
            self.eventId = eventId
            self.date = date
            self.club = club
            self.sessionId = sessionId
            self.sessionLabel = sessionLabel
            self.lapNum = lapNum
            self.timeMs = timeMs
            self.chIdx = chIdx
        }
    }

    /// Flatten event details into one pickable list of laps that have channel
    /// data. Keeps the given event order; within an event, session order then
    /// lap order. Laps without a channel entry (hand-added, no distance
    /// window) drop out, exactly as on the web.
    public static func comparableLaps(_ events: [EventLaps]) -> [Row] {
        var rows: [Row] = []
        for event in events {
            for session in event.sessions {
                guard let channels = session.channels, !channels.laps.isEmpty, !session.laps.isEmpty else {
                    continue
                }
                for match in ChannelGraphs.matchLapsToChannels(session.laps, channels.laps)
                where match.hasChannels {
                    rows.append(
                        Row(
                            eventId: event.eventId,
                            date: event.date,
                            club: event.club,
                            sessionId: session.sessionId,
                            sessionLabel: session.label,
                            lapNum: match.lap.lapNum,
                            timeMs: match.lap.timeMs,
                            chIdx: match.chIdx
                        )
                    )
                }
            }
        }
        return rows
    }

    /// Default selection: side A is "current me" — the best lap of the most
    /// recent event with comparable laps — and side B is "best me", the
    /// overall best lap; when they coincide, B falls back to the best of the
    /// rest. Indexes into the rows, nil when there is nothing to compare.
    public static func defaultComparePicks(_ rows: [Row]) -> (a: Int, b: Int)? {
        guard rows.count >= 2 else { return nil }
        let latest = rows.map(\.date).max() ?? rows[0].date
        func bestIn(_ idxs: [Int]) -> Int {
            idxs.dropFirst().reduce(idxs[0]) { rows[$1].timeMs < rows[$0].timeMs ? $1 : $0 }
        }
        let all = Array(rows.indices)
        let a = bestIn(all.filter { rows[$0].date == latest })
        var b = bestIn(all)
        if b == a { b = bestIn(all.filter { $0 != a }) }
        return (a, b)
    }

    /// Linear-resample one stored channel-lap entry from its grid spacing onto
    /// another. Identity when the spacings already match (every writer uses
    /// 20 m today, but `dStepM` is stored per session and this must not
    /// assume). Same arithmetic as the JS, so the fixture pins every value.
    public static func resampleChannelLap(_ entry: LapChannels, _ fromStepM: Double, _ toStepM: Double) -> LapChannels {
        guard fromStepM != toStepM else { return entry }
        func resample(_ arr: [Double]?) -> [Double]? {
            guard let arr, arr.count >= 2 else { return nil }
            let n = Int((Double(arr.count - 1) * fromStepM / toStepM).rounded(.down)) + 1
            return (0..<n).map { k in
                let p = Double(k) * toStepM / fromStepM
                let i0 = Swift.min(arr.count - 2, Int(p.rounded(.down)))
                return arr[i0] + (arr[i0 + 1] - arr[i0]) * (p - Double(i0))
            }
        }
        return LapChannels(
            n: entry.n,
            timeMs: entry.timeMs,
            speed: resample(entry.speed),
            rpm: resample(entry.rpm),
            latG: resample(entry.latG),
            throttle: resample(entry.throttle),
            brake: resample(entry.brake),
            steering: resample(entry.steering)
        )
    }

    /// Put a pair of channel-lap entries on one grid: side B is resampled onto
    /// side A's spacing when the two sessions stored different grids.
    public static func alignLapPair(
        _ entryA: LapChannels, _ stepA: Double, _ entryB: LapChannels, _ stepB: Double
    ) -> SessionChannels {
        SessionChannels(v: 1, dStepM: stepA, laps: [entryA, resampleChannelLap(entryB, stepB, stepA)])
    }

    /// Driven length a stored entry covers (its grid extent, metres).
    public static func drivenLengthM(_ entry: LapChannels, _ stepM: Double) -> Double {
        let n = entry.speed?.count ?? 0
        return n > 1 ? Double(n - 1) * stepM : 0
    }

    /// Relative driven-length difference between two entries (0 = identical).
    /// Distance is measured from each lap's own start line, so a large ratio
    /// means a different layout, picked line, or off-track excursion — the
    /// comparison still renders, with a warning past `LENGTH_MISMATCH_WARN`.
    public static func lengthMismatchRatio(
        _ entryA: LapChannels, _ stepA: Double, _ entryB: LapChannels, _ stepB: Double
    ) -> Double {
        let la = drivenLengthM(entryA, stepA)
        let lb = drivenLengthM(entryB, stepB)
        let longest = Swift.max(la, lb)
        return longest > 0 ? abs(la - lb) / longest : 0
    }

    /// One lap's channels reduced to head-to-head numbers. Speed stays km/h as
    /// stored (the screen converts for display); percentages are shares of
    /// grid samples — on a uniform distance grid, shares of driven distance.
    public struct Metrics: Sendable, Equatable {
        public var timeMs: Int
        public var topSpeedKph: Double?
        public var minSpeedKph: Double?
        public var avgSpeedKph: Double?
        public var maxRpm: Double?
        public var maxLatG: Double?
        public var fullThrottlePct: Double?
        public var brakingPct: Double?
    }

    public static func lapMetrics(_ entry: LapChannels) -> Metrics {
        let speed = (entry.speed?.isEmpty == false) ? entry.speed : nil
        func share(_ arr: [Double]?, _ min: Double) -> Double? {
            guard let arr, !arr.isEmpty else { return nil }
            return 100 * Double(arr.filter { $0 >= min }.count) / Double(arr.count)
        }
        func max(_ arr: [Double]?) -> Double? {
            guard let arr, !arr.isEmpty else { return nil }
            return arr.max()
        }
        return Metrics(
            timeMs: entry.timeMs,
            topSpeedKph: speed?.max(),
            minSpeedKph: speed?.min(),
            avgSpeedKph: speed.map { $0.reduce(0, +) / Double($0.count) },
            maxRpm: max(entry.rpm),
            maxLatG: max(entry.latG),
            fullThrottlePct: share(entry.throttle, FULL_THROTTLE_PCT),
            brakingPct: share(entry.brake, BRAKING_PCT)
        )
    }
}
